[Feature] Added a Custom Titlebar

ParamBirje opened the pull request:
# ✨ Pull Request

### 📓 Referenced Issue

Fixes #189 

### ℹ️ About the PR

As suggested by @Kidcredo , The [custom-electron-toolbar](https://www.npmjs.com/package/custom-electron-titlebar) package was added to the project to tweak the default toolbar. This made it possible to make the File menu and the Title be in the same space which optimizes the screen space available and more accessible for devices with smaller screen. 

The app's brand color is now used by the custom titlebar.

### 🖼️ Testing Scenarios / Screenshots

Previous titlebar

![previous_titlebar](https://github.com/responsively-org/responsively-app/assets/87022870/4ba9c5e9-5f05-4462-b0c1-dd872e03adb3)

Custom titlebar

![new_titlebar](https://user-images.githubusercontent.com/87022870/275815619-d1cca220-fa47-4772-8d06-bfb004120152.png)

manojVivek commented:
@ParamBirje Sorry for the delayed response, can you please try some subtle color for the background instead of the brand green? 
Something default matching the windows theme should be perfect.

ParamBirje commented:
Hey @manojVivek , I have made the required changes to the titlebar making it's background color subtle and feel default. Here's how it looks.

![image](https://github.com/responsively-org/responsively-app/assets/87022870/ae824137-fb34-41cf-9fb1-680a7040cda4)

manojVivek commented:
Looks goods, also, is it possible to make this a user preference? 

So that it doesn't affect users who like it the current way.

ParamBirje commented:
The custom titlebar actually builds itself during the creation of BrowserWindow, I don't think we can change it in realtime. We could probably save the user preference and on the next restart, load the preferred titlebar. I'll open a new issue/feature for changing the titlebar preferences, if that sounds good?

manojVivek commented:
Yes, applying the change on restart sounds good (we can add a not below the user preference option indicating that). 

Yes, a new issue is fine, but I would like to merge both of these together, as we wouldn't want to surprise the users with something they didn't expect.

ParamBirje commented:
Could you please give me some pointers to how the storage part be implemented? We can't use `window.electron.store` because the window doesn't exist yet. Should it be done using the file system (fs) API in electron or something else?

manojVivek commented:
@ParamBirje You don't need any extra package for this. 

You can directly access the user preferences in the main process like this: https://github.com/responsively-org/responsively-app/blob/34c3b903f7e843413e403cdb35be96697a96d945/desktop-app/src/main/main.ts#L261

ParamBirje commented:
Hey @manojVivek ,

Thank you for the assist on electron store, saved a lot of time. Coincidently the package that I had found was `electron-store` itself, and then noticed it was already implemented in the project. The Titlebar Switch feature which switches between the new custom titlebar and the default titlebar has been implemented.

Custom Titlebar (default preference)

![custom_titlebar](https://github.com/responsively-org/responsively-app/assets/87022870/f03261ce-bf9d-4099-ab03-30c252eba0df)

Default Titlebar

![default_titlebar](https://github.com/responsively-org/responsively-app/assets/87022870/047ada15-c208-48ed-8dcf-c55c3d9b0a1f)

Let me know if there are any changes required.

manojVivek commented on desktop-app/src/main/main.ts:
```diff
@@ -119,7 +125,9 @@ const createWindow = async () => {
     icon: getAssetPath('icon.png'),
+    titleBarStyle: 'hidden',
     webPreferences: {
+      sandbox: false,
```
Why do we want to disable the `sandbox` option?

ParamBirje replied:
The `custom-electron-titlebar` mentioned it in their [documentation](https://github.com/AlexTorresDev/custom-electron-titlebar/wiki/Get-Started), for the library to function properly, `sandbox` should be `false`.

However, I tested the functionality by **omitting** the `sandbox` line, and found out, the custom titlebar works just fine and encountered no issues yet. 

Should I keep this line anyways?

manojVivek replied:
If it works without this, then I would suggest we remove this, as this is option is a compromise on security.

manojVivek commented on desktop-app/src/main/main.ts:
```diff
@@ -119,7 +125,9 @@ const createWindow = async () => {
     width,
     height,
     icon: getAssetPath('icon.png'),
+    titleBarStyle: 'hidden',
```
Also, we want to ensure this doesn't affect Mac and Linux versions of the app, do you have any ideas how it works there after this change?

ParamBirje replied:
I don't think this would affect Mac/Linux versions, as that option disables default taskbar styles and results in a full-size content window.

manojVivek replied:
I just tried this code on Mac and it affects the title bar color.

Preference disabled(default):
<img width="1307" alt="Screenshot 2023-10-29 at 11 39 21 AM" src="https://github.com/responsively-org/responsively-app/assets/1283424/ed1286d9-ecea-471d-bda5-6be58e75d79a">

Preference enabled:
<img width="1193" alt="Screenshot 2023-10-29 at 11 38 42 AM" src="https://github.com/responsively-org/responsively-app/assets/1283424/f45d1e51-49cd-433f-8968-5a04e673e13e">


Is it possible to selectively apply this by checking, only when the platform is windows?

manojVivek replied:
Try moving the styles to a new css file and optionally import based on UA data?
<img width="628" alt="Screenshot 2023-10-29 at 11 53 53 AM" src="https://github.com/responsively-org/responsively-app/assets/1283424/287ca128-7dac-4e72-a018-e275ce75e890">

ParamBirje commented:
Hey @manojVivek ,

Here's the toggle implemented in the Settings panel.

![toggle_in_settings](https://github.com/responsively-org/responsively-app/assets/87022870/ac17c09e-111f-4b95-a4e0-2f18480d4e15)

ParamBirje commented:
Hey @manojVivek ,

I have made changes to the custom titlebar's implementation such that only if the platform is Windows, the feature can be used. The UI which toggles the custom titlebar in the Settings, is now only visible to Windows users. I have also moved the custom titlebar CSS to a new file as directed.

Here's how I am checking platforms in `main.ts` and `preload.ts`

![preload_main](https://github.com/responsively-org/responsively-app/assets/87022870/21cb0ab7-c699-4dca-8c6e-658e148a736c)

And in the React files (process is not exposed here)

![tsx_files](https://github.com/responsively-org/responsively-app/assets/87022870/8a7872d5-5e79-45b8-9064-155794676cec)

[The `sandbox: false` is now removed.]

manojVivek reviewed (approved):
This is looking good! Thanks for patiently addressing all the comments! ♥️

ParamBirje commented:
It's been great dealing with issues and solutions back and forth. Thank you for letting me have a chance at contributing to a great project, with you and fellow maintainers doing a great job at keeping it that way! 😄
